src/blob.jl

```diff
@@ -80,19 +80,45 @@ Defaults to `sizeof(T)`.
     end
 end
 
+# TODO: Sad, none of these below attempts will constant-fold.
+# function self_size(::Type{T}) where T
+#     # @assert isconcretetype(T)
+#     if isempty(fieldnames(T))
+#         sizeof(T)
+#     else
+#         sizes = NTuple{fieldcount(T), Int}(self_size(typ) for typ in fieldtypes(T))
+#         # sizes = self_size.(fieldtypes(T))
+#         sum(sizes)
+#         # _sum_field_sizes(T)
+#     end
+# end
+# @inline _sum_field_sizes(::Type{T}) where {T} = _sum_field_sizes(T, Val(fieldcount(T)))
+# @inline _sum_field_sizes(::Type, ::Val{0}) = 0
+# @inline function _sum_field_sizes(::Type{T}, ::Val{i}) where {T,i}
+#     return self_size(fieldtype(T, i)) + _sum_field_sizes(T, Val(i-1))
+# end
```

Comment:
I'm guessing the issues with this version were in `self_size` rather than `_sum_field_sizes`?

Reply:
not sure. I think it might be the mutual recursion between the two functions?

It ends up fully *type stable*, but not fully **const-folded** like it was before.

With the generated function, the `@code_typed` is literally just `return 49`. But with this one, it performs the computation at runtime still.

Reply:
This one constant folds for the `Bar` struct in the tests:
```julia
Base.@assume_effects :total function self_size(::Type{T}) where T
    if isempty(fieldnames(T))
        sizeof(T)
    else
        +(Iterators.map(self_size, fieldtypes(T))...)
    end
end

julia> Blobs.self_size(Bar)
49

julia> @code_llvm Blobs.self_size(Bar)
;  @ /Users/tdrvostep/.julia/dev/Blobs/src/blob.jl:83 within `self_size`
define i64 @julia_self_size_444() #0 {
top:
  ret i64 49
}
```

The following also constant folds:
```julia
Base.@assume_effects :total !:nothrow !:terminates_locally !:notaskstate !:inaccessiblememonly function self_size(::Type{T}) where T
    if isempty(fieldnames(T))
        sizeof(T)
    else
        +(Iterators.map(self_size, fieldtypes(T))...)
    end
end
```
So, looking at the docs for `:total` in `Base.@assume_effects` I thought this would constant-fold as well, but it doesn't:
```julia
Base.@assume_effects :consistent :effect_free function self_size(::Type{T}) where T
    if isempty(fieldnames(T))
        sizeof(T)
    else
        +(Iterators.map(self_size, fieldtypes(T))...)
    end
end
```

Reply:
Oh wow, @Drvi, _amazing!_ ❤️ 

Wanna commit these and push them up?? Feel free to take over the PR or join in on it!

Reply:
>So, looking at the docs for :total in Base.@assume_effects I thought this would constant-fold as well, but it doesn't.

It looks like you also need `:terminates_globally` to constant-fold, which you can also get by setting `:foldable`.

```
  :foldable
  =========

  This setting is a convenient shortcut for the set of effects that the compiler requires to be guaranteed to constant fold a call at
  compile time. It is currently equivalent to the following settings:

    •  :consistent
    •  :effect_free
    •  :terminates_globally

  │ Note
  │
  │  This list in particular does not include :nothrow. The compiler will still attempt constant propagation and note any thrown
  │  error at compile time. Note however, that by the :consistent-cy requirements, any such annotated call must consistently throw
  │  given the same argument values.

  │ Note
  │
  │  An explicit @inbounds annotation inside the function will also disable constant folding and not be overriden by :foldable.
```
And indeed, this folds:
```julia
Base.@assume_effects :foldable function self_size(::Type{T}) where T
    # This function is marked :total to encourage constant folding for this types-only
    # static computation.
    if isempty(fieldnames(T))
        sizeof(T)
    else
        +(Iterators.map(self_size, fieldtypes(T))...)
    end
end
```

It does seem to actually _be_ total from what I could read, but the docstring says to try to stick to the smallest possible assertion, so i think we can stick with `:foldable`:
> Even in such a case, you SHOULD take all possible steps to minimize the strength of the effect assertion (e.g., do not use :total if :nothrow would have been sufficient).

Reply:
Pushed up a commit with `:foldable`, and recursion, after benchmarking. Thanks @Drvi and @Sacha0!